File: src/main/scala/resonant/lib/grid/GridNode.java
package resonant.lib.grid;

import resonant.api.grid.INodeGrid;
import resonant.lib.grid.node.Node;

/**
 * Created by robert on 8/13/2014.
 */
public class GridNode<N extends Node> extends Grid<N>
{
    public GridNode(Class node)
    {
        super(node);
    }

    @Override
    protected void reconstructNode(N node)
    {
        node.reconstruct();
        if(node instanceof INodeGrid)
            ((INodeGrid)node).setGrid(this);
    }

    @Override
    public void deconstruct()
    {
        for(N node : getNodes())
        {
            if(node instanceof INodeGrid)
                ((INodeGrid)node).setGrid(null);
            node.reconstruct();
        }
        super.deconstruct();
    }
}

File: src/main/scala/resonant/lib/grid/GridNode.scala
package resonant.lib.grid

import resonant.lib.grid.node.NodeGrid

import scala.collection.convert.wrapAll._

/**
 * A grid that contains nodes where all nodes are interconnected
 *
 * @param N The type of node we can connect
 * @author Calclavia
 */
class GridNode[N <: NodeGrid[N]](node: Class[N]) extends Grid[N](node)
{
  /**
   * Rebuild the node list starting from the first node and recursively iterating through its connections.
   */
  def reconstruct(first: N)
  {
    if (!getNodes.contains(first))
    {
      //TODO: Reconstruct may be called MANY times unnecessarily multiple times. Add check to prevent extra calls
      getNodes.clear()
      populate(first)
    }
  }

  /**
   * Populates the node list
   */
  protected def populate(node: N, prev: N = null.asInstanceOf[N])
  {
    if (!getNodes.contains(node) && isValidNode(node))
    {
      add(node)
      reconstructNode(node)
      node.connections.foreach(n => populate(n, node))
    }
  }

  protected def populateNode(node: N)
  {
    node.setGrid(this)
  }

  override def deconstruct()
  {
    for (node <- getNodes)
    {
      node.setGrid(null)
    }

    super.deconstruct()
  }
}